posix: signal: APIs implementation

ycsin opened the pull request:
Implementation for the following POSIX signal APIs:

- [`int sigemptyset(sigset_t *set);`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sigemptyset.html#)
- [`int sigfillset(sigset_t *set);`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sigfillset.html#)
- [`int sigaddset(sigset_t *set, int signo);`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sigaddset.html#)
- [`int sigdelset(sigset_t *set, int signo);`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sigdelset.html#)
- [`int sigismember(const sigset_t *set, int signo);`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sigismember.html#)
- [`char *strsignal(int signum);`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/strsignal.html#)

fixes #59929
fixes #59931
fixes #59927
fixes #59928
fixes #59932

ycsin commented on tests/posix/common/src/signal.c:
```diff
@@ -0,0 +1,241 @@
+	zassert_equal(errno, EINVAL, "errno should be %s", "EINVAL");
+}
+
+#define SIG_WD_IDX(_signo)  ((int)((_signo - 1) / BITS_PER_LONG))
+#define SIG_MSK_SET(_signo) (WRITE_BIT(sig_mask[SIG_WD_IDX(_signo)], (_signo - 1) % _NSIG_BPW, 1))
```
need to update these, end result should be similar to `test_signal_ismember`

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,60 @@
+
+typedef struct {
+	unsigned long sig[_NSIG_WORDS];
+} sigset_t;
```
we should disable this in `scripts/checkpatch.pl` for posix headers, it's a false positive.

cfriedt commented on lib/posix/signal.c:
```diff
@@ -0,0 +1,73 @@
+
+#include <zephyr/posix/signal.h>
+
+static inline int signo_word_idx(int signo)
```
static inline might not be correct here.

ycsin replied:
converted to macro

cfriedt commented on tests/posix/common/src/signal.c:
```diff
@@ -0,0 +1,241 @@
+
+	WRITE_BIT(sig_mask, SIGHUP - 1, 0);
+	zassert_ok(sigdelset(&set, SIGHUP));
+	zassert_equal(set.sig[0], sig_mask, "set.sig[%d of %d] has content: %lx, expected %lx", 0,
```
this test is failing
```
twister -i -p qemu_arc_hs6x   -s tests/posix/common/portability.posix.common.tls
```
https://github.com/zephyrproject-rtos/zephyr/actions/runs/5478149532/jobs/9978281717?pr=60083#step:13:765

ycsin committed:
checkpatch: exclude POSIX headers from typedef rule

The `NEW_TYPEDEFS` rule consistently generates false
positives when implementing POSIX standard APIs.

It makes sense to disable this check for the POSIX haders
rather than requiring merge superpowers constantly. That
way, we can merge as per usual after sufficient approvals
rather than waiting for someone with merge superpowers to
override / manually merge.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,60 @@
+#ifdef CONFIG_POSIX_SIGNAL
+#define _NSIG       64
+#define _NSIG_BPW   BITS_PER_LONG
+#define _NSIG_WORDS (_NSIG / _NSIG_BPW)
```
used `DIV_ROUND_UP`

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,61 @@
 
+#ifdef CONFIG_POSIX_SIGNAL
+#define _NSIG       64
+#define _NSIG_BPW   BITS_PER_LONG
```
_NSIG_BPW can probably just be replaced with BITS_PER_LONG

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,61 @@
+#ifdef CONFIG_POSIX_SIGNAL
+#define _NSIG       64
+#define _NSIG_BPW   BITS_PER_LONG
+#define _NSIG_WORDS (DIV_ROUND_UP(_NSIG, _NSIG_BPW))
```
I would move this define into signal.c and use the expanded form in the header

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,61 @@
 #endif
 
+#ifdef CONFIG_POSIX_SIGNAL
+#define _NSIG       64
```
Instead of _NSIG hard-coded can we use a Kconfig? Be sure to build assert that it is greater than or equal to sig rt min

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,61 @@
+#define SIGRTMAX _NSIG
+#endif
+
+typedef struct {
```
All structs, typedefs and functions should be grouped together in this header below macros .

cfriedt commented on lib/posix/strsignal.c:
```diff
@@ -0,0 +1,93 @@
+		return "Invalid signal";
+	}
+
+	switch (signum) {
```
This seems kind of clunky. We should be able to use a LUT. Can you check out what I did for strerror in the minimal libc? If you add the string beside the define as a comment, then you can derive the table at build time.

ycsin replied:
do you mean with a [script](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/gen_strerror_table.py) to generate that?

cfriedt commented on tests/posix/common/src/signal.c:
```diff
@@ -0,0 +1,251 @@
+
+ZTEST(posix_apis, test_signal_addset_oor)
+{
+	int rc;
```
Can probably get away without rc

ycsin replied:
fixed others as well

cfriedt commented on tests/posix/common/src/signal.c:
```diff
@@ -0,0 +1,251 @@
+	sigset_t set;
+	sigset_t target;
+
+	memset(&set, 0, sizeof(set));
+	memset(&target, 0, sizeof(target));
```
```suggestion
	set = (sigset_t){0};
	target = (sigset_t){0};
```

cfriedt commented on tests/posix/common/src/signal.c:
```diff
@@ -0,0 +1,251 @@
+
+	signo = SIGRTMIN;
+	zassert_ok(sigaddset(&set, SIGRTMIN));
+	WRITE_BIT(target.sig[0], signo - 1, 1);
```
we can spare an extra bit. I.e. let's use `signo` instead of `signo-1`

cfriedt commented on lib/posix/signal.c:
```diff
@@ -0,0 +1,62 @@
+
+#include <zephyr/posix/signal.h>
+
+#define SIGNO_WORD_IDX(_signo) ((signo - 1) / _NSIG_BPW)
+#define SIGNO_WORD_BIT(_signo) ((signo - 1) % _NSIG_BPW)
```
```suggestion
#define SIGNO_WORD_IDX(_signo) (signo / BITS_PER_LONG)
#define SIGNO_WORD_BIT(_signo) (signo & BIT_MASK(BITS_PER_LONG))
```

cfriedt replied:
I think the latter `&` might be wrong - it should be something like `BIT_MASK(LOG2(BITS_PER_LONG))`

cfriedt commented on lib/posix/signal.c:
```diff
@@ -0,0 +1,62 @@
+
+int sigemptyset(sigset_t *set)
+{
+	memset(set, 0, sizeof(*set));
```
```suggestion
	*set = (sigset_t){0};
```

cfriedt commented on lib/posix/signal.c:
```diff
@@ -0,0 +1,62 @@
+
+int sigfillset(sigset_t *set)
+{
+	memset(set, -1, sizeof(*set));
```
memset is byte oriented while the set array is word oriented. A loop here would likely be faster than memset (not that it's a huge gain)

cfriedt commented on lib/posix/signal.c:
```diff
@@ -0,0 +1,62 @@
+
+int sigaddset(sigset_t *set, int signo)
+{
+	if (!SIGNO_VALID(signo)) {
```
We can probably get away with a static inline here instead of a macro.

cfriedt commented on lib/posix/signal.h:
```diff
@@ -0,0 +1,17 @@
+#include <zephyr/posix/signal.h>
+
+/* Check if signal is valid */
+#define SIGNO_VALID(_signo) ((_signo > 0) && (_signo <= _NSIG))
```
I would put all of this into signal.c

cfriedt commented on lib/posix/strsignal.c:
```diff
@@ -0,0 +1,93 @@
+#define SIG_STRING(_signo, _desc) _signo
+#endif
+
+char *strsignal(int signum)
```
I would put all of this into signal.c too

cfriedt commented on tests/posix/common/src/signal.c:
```diff
@@ -0,0 +1,251 @@
+#include <errno.h>
+#include <signal.h>
+
+#include <zephyr/ztest.h>
```
Please try to make headers alphabetically stored within their groups.

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,62 @@
+#define SIGUNUSED 31
+
+#define SIGRTMIN 32
+#ifndef SIGRTMAX
```
This conditional is a bit questionable - does anything break if we removed it and just define `SIGRTMAX`?

ycsin replied:
Probably won't break anything, can move into Kconfig

ycsin replied:
Probably won't break anything, can probably move into Kconfig as well

cfriedt commented:
Looks like a flakey test. Please fix in a separate commit (should probably be moved down the stack).
```
2023-07-08T08:36:34.3596090Z FAILED: CMakeFiles/app.dir/src/signal_h.c.obj 
2023-07-08T08:36:34.3603241Z ccache /opt/toolchains/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DCORE_CM4 -DHSE_VALUE=24000000 -DKERNEL -DPICOLIBC_INTEGER_PRINTF_SCANF -DSTM32F427xx -DTC_RUNID=926c49688999a665fa9f7ab0c9114eb6 -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=1 -D_POSIX_C_SOURCE=200809 -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I/__w/zephyr/zephyr/include -I/__w/zephyr/zephyr/twister-out/96b_aerocore2/tests/posix/headers/portability.posix.headers.picolibc.with_posix_api/zephyr/include/generated -I/__w/zephyr/zephyr/soc/arm/st_stm32/stm32f4 -I/__w/zephyr/zephyr/include/zephyr/posix -I/__w/zephyr/zephyr/lib/posix/getopt/. -I/__w/zephyr/zephyr/drivers -I/__w/zephyr/zephyr/soc/arm/st_stm32/common -I/__w/zephyr/zephyr/subsys/testsuite/include -I/__w/zephyr/zephyr/subsys/testsuite/ztest/include -I/__w/zephyr/zephyr/subsys/net/lib/sockets/. -I/__w/zephyr/zephyr/subsys/net/lib/dns/. -I/__w/zephyr/modules/lib/acpica/../../zephyr/include -I/__w/zephyr/modules/lib/acpica/.. -I/__w/zephyr/modules/lib/acpica/source/include -I/__w/zephyr/modules/lib/acpica/source/include/platform -I/__w/zephyr/modules/lib/acpica/source/compiler -I/__w/zephyr/modules/lib/acpica/generate/zephyr -I/__w/zephyr/modules/lib/acpica/source/tools/acpiexec -I/__w/zephyr/modules/lib/acpica/source/tools/acpidump -I/__w/zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/__w/zephyr/modules/hal/stm32/stm32cube/stm32f4xx/soc -I/__w/zephyr/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include -I/__w/zephyr/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include/Legacy -I/__w/zephyr/modules/hal/stm32/stm32cube/common_ll/include -fno-strict-aliasing -Werror -Os -imacros /__w/zephyr/zephyr/twister-out/96b_aerocore2/tests/posix/headers/portability.posix.headers.picolibc.with_posix_api/zephyr/include/generated/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee --sysroot=/opt/toolchains/zephyr-sdk-0.16.1/arm-zephyr-eabi/arm-zephyr-eabi -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/tests/posix/headers=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections --specs=picolibc.specs -D_POSIX_THREADS -std=c99 -MD -MT CMakeFiles/app.dir/src/signal_h.c.obj -MF CMakeFiles/app.dir/src/signal_h.c.obj.d -o CMakeFiles/app.dir/src/signal_h.c.obj -c /__w/zephyr/zephyr/tests/posix/headers/src/signal_h.c
2023-07-08T08:36:34.3608355Z /__w/zephyr/zephyr/tests/posix/headers/src/signal_h.c: In function 'posix_headers_test_signal_h':
2023-07-08T08:36:34.3609245Z /__w/zephyr/zephyr/tests/posix/headers/src/signal_h.c:28:9: error: conversion to non-scalar type requested
2023-07-08T08:36:34.3609763Z    28 |         zassert_not_equal((sigset_t)-1, (sigset_t)0);
2023-07-08T08:36:34.3610070Z       |         ^~~~~~~~~~~~~~~~~
```

ycsin committed:
posix: signal: add initial header

Initial header for the signal APIs.
APIs to be implemented in later commit.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
posix: signal: implement sigemptyset

Implementation and ztest for sigemptyset.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
posix: signal: implement sigfillset

Implementation and ztest for sigfillset.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
posix: signal: implement sigaddset

Implementation and ztest for sigaddset.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
posix: signal: implement sigdelset

Implementation and ztest for sigdelset.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
posix: signal: implement sigismember

Implementation and ztest for sigismember.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,74 @@
+
+#define _NSIG     CONFIG_POSIX_SIGNAL_NSIG
+
+#define SIGHUP    1
```
It would be nice to re-use `scripts/build/gen_strerror_table.py` a bit the same way it's used in `lib/libc/minimal/CMakeLists.txt`.

In this file, e.g.
```cpp
#define SIGHUP 1 /* Hangup */
```

ycsin replied:
most of the differences between `strerror` & `strsignal` can be solved by adding more options to the script, but I'm not sure what to do about the preprocessor defines of `strerror` so I've opted to create another script for `strsignal` instead of modifying `gen_strerror_table.py`, let me know what you  think.

cfriedt replied:
Would be nice to eventually merge the two, but no pressure yet.

cfriedt commented on lib/posix/Kconfig.signal:
```diff
@@ -0,0 +1,37 @@
+
+config POSIX_SIGNAL_SIGRTMIN
+	int "Minimum number of real-time signal"
+	default 32
```
Hmm.. this might be better off as hard-coded, i.e. without a Kconfig

cfriedt commented on lib/posix/posix_internal.h:
```diff
@@ -42,6 +42,9 @@ struct posix_thread {
 
 	/* Queue ID (internal-only) */
 	uint8_t qid;
+
```
This can probably be removed until e.g. `pthread_kill()` and `sigsetmask()` are implemented.

cfriedt commented on lib/posix/signal.c:
```diff
@@ -0,0 +1,119 @@
+	static char other_sigstr[sizeof("RT signal xx")];
+
+	if (!signo_valid(signum)) {
+		return "Invalid signal";
```
You can probably set errno to EINVAL here
https://pubs.opengroup.org/onlinepubs/9699919799/functions/strsignal.html

cfriedt commented on lib/posix/signal.c:
```diff
@@ -0,0 +1,119 @@
+	return 1 & (set->sig[SIGNO_WORD_IDX(signo)] >> SIGNO_WORD_BIT(signo));
+}
+
+char *strsignal(int signum)
```
if CONFIG_POSIX_SIGNAL_STRING_FULL is not set, then you can just return the empty string from this function. I.e. keep the function sort-of functional but do not use any extra ROM.

As long as you wrap it in an `if (!IS_ENABLED())` and make that happen early, the table should be dead-stripped.

cfriedt commented on lib/posix/signal.c:
```diff
@@ -0,0 +1,119 @@
+	return other_sigstr;
+}
+
+int sigpending(sigset_t *set)
```
This is actually part of a separate subprofiling option and is a bit more complicated.

I would suggest dropping this from your current changeset. It will require some special attention

cfriedt commented on lib/posix/signal_internal.h:
```diff
@@ -0,0 +1,20 @@
+#include <zephyr/posix/signal.h>
+#include <zephyr/sys/slist.h>
+
+struct sig_pending {
```
I would suggest dropping this from your current changeset.

cfriedt commented on scripts/build/gen_strsignal_table.py:
```diff
@@ -0,0 +1,103 @@
+        dest='output',
+        required=True,
+        help='output file (e.g. build/zephyr/misc/generated/lib/posix/strsignal_table.h)')
+    parser.add_argument(
```
Can we simplify this for now by simply always using the full value in the table?

The other space optimization is to return an empty string from strsignal when the table option is not selected

cfriedt commented on tests/posix/common/src/signal.c:
```diff
@@ -0,0 +1,270 @@
+#endif
+}
+
+static void *sigpending_test_fn(void *arg)
```
It's probably better to drop sigpending from your current changeset.

ycsin committed:
posix: signal: implement strsignal

Implementation and ztest for strsignal.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt reviewed (dismissed):
Good enough for now ;-)

ycsin commented:
remove one unused line in the CMakeLists.txt

keith-packard commented:
Reviewing the functions provided, what's the use case for this functionality? Are there plans to extend this to provide `signal` and `raise`? How about the pthread signal functions like `pthread_kill`? Given how difficult most developers find this particular programming model, I could build a fairly strong argument for just not providing any of this functionality and having developers use other communication mechanisms instead ...

cfriedt commented:
> Reviewing the functions provided, what's the use case for this functionality? Are there plans to extend this to provide `signal` and `raise`?

Currently, we have no plans to (fully) support the `POSIX_MULTI_PROCESS` subprofiling option, since Zephyr hasn't really ventured into that space yet, so probably not `signal()` and `raise()` (maybe only if there is a certain use case for it).

> How about the pthread signal functions like `pthread_kill`?

Yes `pthread_kill()` and `pthread_sigmask()` are the only things I can see working in a single-process environment. 

> Given how difficult most developers find this particular programming model, I could build a fairly strong argument for just not providing any of this functionality and having developers use other communication mechanisms instead ...

It's definitely easier to use other communication mechanisms, and so this will be defaulted to N, of course. It's mainly going to be there to have something for external libraries to integrate with.

One particular thing to note is that we will likely never interrupt any system call. So there is a bit of wiggle room in the mechanism for "delivering" signals even.

keith-packard commented:
> Currently, we have no plans to (fully) support the `POSIX_MULTI_PROCESS` subprofiling option, since Zephyr hasn't really ventured into that space yet, so probably not `signal()` and `raise()` (maybe only if there is a certain use case for it). Some parts of it are already supported (e.g. assert, exit, sleep, exit, ..)

There's a whole mess of interrelated functions in the C spec (these aren't  POSIX, after all). `assert` is defined to call `abort`, `abort` is defined to call `raise(SIGABRT)`, and that requires that `signal` be available to specify a handler. These all work fine in a single process environment; the C standard doesn't even mention the `kill` function, after all. 

Now, whether you want to follow the C spec for `assert` is a whole 'nuther story; I've never heard of anyone trapping `SIGABRT` and doing anything with the result, but I can't say it couldn't happen. And, in an embedded system, it's actually pretty plausible for the application to want to perform some specific logging and error recovery when assertions fail, so maybe this whole thing might actually be useful to someone?

> It's definitely easier to use other communication mechanisms, and so this will be defaulted to N, of course. It's mainly going to be there to have something for external libraries to link with.

Yeah, makes sense. One suspects those will call `signal` a bunch of times to ignore fun POSIX things like SIGTSTP, and maybe mess with sigmasks just for fun.

> One particular thing to note is that we will likely never interrupt any system call. So there is a bit of wiggle room in the mechanism for "delivering" signals even.

Given that you aren't even defining an OS-delivered signals, or the `kill` function, I don't think there's any issue here. Consider whether you want to support the C standard interpretation of signal handling and how much of the fancier POSIX stuff you want to bolt on as well.

cfriedt commented:
@keith-packard - if we're going to tackle the ISO C variants of signal, etc, it will likely be later. Potentially some interesting implications though, for sure.

Are you ok to approve this one?

keith-packard commented on lib/posix/signal.c:
```diff
@@ -0,0 +1,99 @@
+
+char *strsignal(int signum)
+{
+	static char buf[sizeof("RT signal xx")];
```
This tacitly assumes that _NSIG < 132, I think?

ycsin replied:
I used

```c
char buf[sizeof("RT signal " STRINGIFY(SIGRTMAX))] = {0};
```
 in the test and I forgot to update it here, let me do that real quick

ycsin replied:
fixed this to match with the test

keith-packard reviewed (dismissed):
Aside from `strsignal` assuming that there won't be more than 99 RT signals, this all looks fine.

ycsin commented:
fixed a typo in the `sigismember` test

ycsin commented:
another issue here is that the tests are hardcoded to support only 63 signals, my intention was to keep the test straightforward and use as least algo to verify the APIs as possible, but since `POSIX_SIGNAL_NSIG` is configurable, this might not make sense, adding DNM for now

ycsin committed:
test: posix: signal: add test for big nsig number

Add an extra test config to test large number of signals.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
- fixed a bug in how the number `sig` array is declared
- fixed the test to allow probably any number of signal
- added a new test config to test large number of signals
- removed `DNM`, should be good to go now

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,62 @@
+/* 30 not used */
+#define SIGSYS    31 /**< Bad system call */
+
+#define SIGRTMIN 32
+#define SIGRTMAX _NSIG
```
Kind of curious - is there a hard requirement (from the spec) that `SIGRTMAX > SIGRTMIN`? I.e. could we set `#define SIGRTMAX (_NSIG - 1)`

cfriedt replied:
Ah, I see - this might help a bit
https://pubs.opengroup.org/onlinepubs/7908799/xsh/limits.h.html

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,62 @@
+/* 30 not used */
+#define SIGSYS    31 /**< Bad system call */
+
+#define SIGRTMIN 32
+#define SIGRTMAX _NSIG
```
```suggestion
#define SIGRTMIN 32
#define SIGRTMAX (SIGRTMIN + CONFIG_POSIX_RTSIG_MAX)
#define _NSIG (SIGRTMAX + 1)
```

cfriedt commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,62 @@
+#define SIGRTMIN 32
+#define SIGRTMAX _NSIG
+
+BUILD_ASSERT(CONFIG_POSIX_SIGNAL_NSIG > SIGRTMIN);
```
```suggestion
BUILD_ASSERT(CONFIG_POSIX_RTSIG_MAX >= 0);
```

ycsin commented:
@keith-packard may I have your review again?

keith-packard commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,57 @@
+#define SIGSYS    31 /**< Bad system call */
+
+#define SIGRTMIN 32
+#define SIGRTMAX (SIGRTMIN + CONFIG_POSIX_RTSIG_MAX)
```
Looks like you should define `RTSIG_MAX` to be `CONFIG_POSIX_RTSIG_MAX` in `<limits.h>` and that it must be at least 8.

`https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html`

ycsin replied:
I think this Kconfig will make the implementation of `limits.h` a bit easier, but I think `limits.h` can probably be in another PR

cfriedt replied:
Yeah, that sounds like it's getting close to `sysconf()`. There is another PR in draft for that, so I can add `RTSIG_MAX` separately

ycsin commented on lib/posix/Kconfig.signal:
```diff
@@ -0,0 +1,25 @@
+	int "Maximum real-time signal number"
+	default 31
+	help
+	  Define the maximum real-time signal number.
```
Updated the desc to align with the [doc](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html):

```
config POSIX_RTSIG_MAX
	int "Maximum number of realtime signals"
	default 31
	help
	  Define the maximum number of realtime signals (RTSIG_MAX).
	  The range of realtime signals is [SIGRTMIN .. (SIGRTMIN+RTSIG_MAX)]
```

keith-packard commented on include/zephyr/posix/signal.h:
```diff
@@ -12,6 +12,57 @@
+#define SIGRTMAX (SIGRTMIN + CONFIG_POSIX_RTSIG_MAX)
+#define _NSIG (SIGRTMAX + 1)
+
+BUILD_ASSERT(CONFIG_POSIX_RTSIG_MAX >= 0);
```
POSIX says 8; that could get checked here or Kconfig, or just stick a note in Kconfig indicating what will make the system POSIX compliant.

ycsin replied:
I've added `Kconfig.limits` to specify the default limit to be posix compliant, but user can freely changes this Kconfig to be anything >= 0, let me know what you think about this approach

These Kconfigs serve as direct translation when we implement `limits.h` in the future

cc @cfriedt

keith-packard replied:
That sounds perfect -- tell people how to make POSIX compliant systems, but also allow them to reduce resources if desired.

keith-packard reviewed (dismissed):
Looking good.

ycsin committed:
posix: signal: check RTSIG limit

Use build assert to make sure that the realtime signal
constants are configured properly in the Kconfig.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt reviewed (approved):
LGTM
